### DISCUSSION TITLE
Viewing Postgres logs?

### DISCUSSION THREAD

**Discussion by emschwartz** [supabase/supabase#479] [Questions] [answered]
Is there a way to view the postgres logs? I'm trying to debug a stored function + trigger but the auth API only says `Database error saving new user`.

If there isn't currently a way to see the logs, this might be a helpful thing to add in the future.

**Comment by madebyfabian** [comment]
I think there is no way to view this currently. But what you can do is of course open the database (with the credentials provided by supabase, e.g. host, port, etc.) with any other tool (for example "heidisql" and then try to execute these. These tools often give you way more accurate error logs/reporting. That's what I did so far when I tried to build some functions and I had no clue why they didn't worked.

But yeah, this feature would be great!

**Reply by kiwicopple** [reply]
Thanks for answering! 

Agreed, it would be awesome from the dashboard - we're working on it :)

**Reply by chipilov** [reply]
I think the logs are now exposed via the Dashboard in Settings -> Logs -> Database.

However, it's not clear to me how to use the "Search logs via query" functionality.

**Reply by Ziinc** [reply]
Hi @emschwartz  @madebyfabian , 
Additional logging features have been made available in the Dashboard, as @chipilov has mentioned. Documentation on its features and capabilities are available [here](https://supabase.com/docs/guides/platform/logs).

**Comment by MohamedLamineAllal** [comment]
any news on this!

I'm running a cron schedule from the query editor

```sql
select
  cron.schedule(
    'check-test-run-2', -- name of the cron job
    '* * * * *', -- every minute
    $$
      insert into temp_test_log (issuer, message)
        values('Cron-job check run webhook', 'POST/ { arn: 1 }')
    $$
  );
```
But no entry is added to the test table!

![image](https://user-images.githubusercontent.com/20062543/128054712-5d2f147a-da6e-43fe-82ab-22b1472514ff.png)

Extension enabled!

The job is added to the cron.job table! But the insert command is not being executing! or Something is not right!

How does heidisql help with a cron job here! Does it access some logs or ??

For permissions! i gave permission to (service_role, anon, supabase_admin, dashboard_user) I know some of them are super user and doesn't require!

Without logs! it's hard to know what's going on!
Any helping elements and idea?

**Comment by MohamedLamineAllal** [comment]
![image](https://user-images.githubusercontent.com/20062543/128097792-afe5184e-dab7-47dc-a8ce-8be04727bb19.png)

Checking the table as mentioned on the article! i found out that the command execution is failing! Because the connection failed!

![image](https://user-images.githubusercontent.com/20062543/128097944-fd28a3fb-59d5-4bfb-9954-2b93dec0c0c7.png)

I made the user a super user temporally! But then it seems not the thing!
And it's something that require Db Authentication config!

https://github.com/citusdata/pg_cron/issues/66

Anything that i can do! I'm inserting a row from nodejs (service_role)! Which trigger a function that schedule the cron! The cron schedule all right and work ! But fail at command execution! What can i do! And what is the detail of the why ?

**Reply by MohamedLamineAllal** [reply]
I ended up creating a total new user (role)! With it's password! And used that for the external connection! I gave the new user the necessary priveleges! And all worked well! (Using service_role was problematic! Even by giving it super user privileges! it didn't work! And fail by a 'connection failed' error!

**Comment by tartard** [comment]
Hello, I just found a solution with the **file_fdw extension** since supabase postgres now logs in CSV file :  
https://www.postgresql.org/docs/13/file-fdw.html
https://www.postgresql.org/docs/13/runtime-config-logging.html#RUNTIME-CONFIG-LOGGING-CSVLOG


To setup the log table , run the following SQL to get the postgres CSV log file path :

```sql
select pg_current_logfile();
``` 
Currently the log path is **'pg_log/postgresql.csv',**
( https://github.com/supabase/postgres/blob/020cef4ba760c65c75a627174579b4d4e43f311c/ansible/files/postgresql_config/postgresql.conf.j2#L443 )  

Then run :
```sql
CREATE EXTENSION file_fdw;

CREATE SERVER logserver FOREIGN DATA WRAPPER file_fdw;

CREATE FOREIGN TABLE pglog (
  log_time timestamp(3) with time zone,
  user_name text,
  database_name text,
  process_id integer,
  connection_from text,
  session_id text,
  session_line_num bigint,
  command_tag text,
  session_start_time timestamp with time zone,
  virtual_transaction_id text,
  transaction_id bigint,
  error_severity text,
  sql_state_code text,
  message text,
  detail text,
  hint text,
  internal_query text,
  internal_query_pos integer,
  context text,
  query text,
  query_pos integer,
  location text,
  application_name text,
  backend_type text
) SERVER logserver
OPTIONS ( filename 'pg_log/postgresql.csv', format 'csv' );
```

Filesystem logrotate is set with the following conf for the postgres logs : 
https://github.com/supabase/postgres/blob/982ef70bc11ff38016fba69f2f6194509e108bfd/ansible/files/logrotate-postgres

**Reply by CR1AT0RS** [reply]
Just a quick update I had to delete the   "backend_type text" as I was getting error missing data.

**Reply by tartard** [reply]
Ah, I tried in an instance running PostreSQL 13. Maybe are you running PostgreSQL 12 ?

**Reply by keybits** [reply]
Thanks for sharing this. For anyone else trying this:

1. `select pg_current_logfile();` gave me `/var/log/postgresql/postgresql.csv` as of today on my project on https://app.supabase.io/
2. You won't see the table in the Supabase ui, but connect with a client and it's there (I can recommend Beekeeper studio)

**Reply by chipilov** [reply]
After the upgrade to PostgreSQL 14, it seems that 2 more fields need to be added to the end of the foreign table column list for the CSV-parsing to work:
```
  leader_pid integer,
  query_id bigint
```

**Comment by Ziinc** [comment, marked as answer]
Hi @emschwartz @madebyfabian , Additional logging features have been made available in the Dashboard, as @chipilov has mentioned. Documentation on its features and capabilities are available [here](https://supabase.com/docs/guides/platform/logs).

**Reply by madebyfabian** [reply]
That’s great! Thanks for the update :)

**Reply by dsebastien** [reply]
Are those logging features supposed to work during local development? I've tried accessing the pgsql logs, but I get the following error:

```
{
  "code": 500,
  "message": "LOGFLARE_API_KEY env variables are not set"
}
```

Prior to the last update of my docker containers, I could access the log files and get what I needed, now I can't since the log file seems to have been disabled.

**Reply by Ziinc** [reply]
This bug should have been patched in the latest cli version.

You can follow the guide to enable logging here https://supabase.com/docs/guides/getting-started/local-development#enabling-local-logging

**Reply by dsebastien** [reply]
Are there plans to remove the need for us to setup Google Cloud Big Query with billing just to have logs during development, while the local pgsql can provide an easy to use alternative? When pgsql logs are enabled, a simple ls + cat against the docker container is enough to go through the logs. It's easier, faster and cheaper

**Reply by Ziinc** [reply]
yes there are plans to remove the external dependency for a more seamless experience, hence why it is an opt-in feature.

**Reply by cdimitroulas** [reply]
Is there a current workaround for local development logs that doesn't rely on Big Query where we can see the logs in the local postgres instance directly somehow?

**Reply by Ziinc** [reply]
You can use Docker Desktop to browse logs by default. Postgres backend is not supported yet by Logflare.

**Reply by hevar** [reply]
@Ziinc where can the logs be found on Docker Desktop?

**Reply by tylerkatz** [reply]
@hevar Run `docker logs -f <container-name>` on the Supabase Postgres container (port 54322)
